File: client/src/modules/assetmanager/schemas/financial/income-statement.schemas.ts
/**
 * Income Statement Schemas
 *
 * Zod validation schemas for IncomeStatement model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/financial_models.py
 * - Schema: /server/apps/assetmanager/schemas/financial_schemas/income_statement_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/financial_subrouters/income_statement_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// IncomeStatement Schema (Full Representation)
// ==========================================

/**
 * IncomeStatement schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class IncomeStatement(BaseModel)
 */
export const IncomeStatementSchema = z.object({
  id: z.number(),
  entity_id: z.number(),

  // Time Dimensions
  year: z.number(),
  quarter: z.string().nullable(),
  semester: z.string().nullable(),
  month: z.string().nullable(),
  full_year: z.boolean(),
  scenario: z.string(),
  period_start: z.string().nullable(),
  period_end: z.string().nullable(),

  // Revenue
  revenue: z.number().nullable(),
  cost_of_goods: z.number().nullable(),
  gross_profit: z.number().nullable(),

  // Operating Expenses
  research_and_development: z.number().nullable(),
  selling_general_and_administrative: z.number().nullable(),
  other_operating_expenses: z.number().nullable(),

  // Results
  operating_income: z.number().nullable(),
  non_operating_interest_income: z.number().nullable(),
  non_operating_interest_expense: z.number().nullable(),
  other_income_expense: z.number().nullable(),
  pretax_income: z.number().nullable(),
  income_tax: z.number().nullable(),
  net_income: z.number().nullable(),

  // Additional
  eps_basic: z.number().nullable(),
  eps_diluted: z.number().nullable(),
  basic_shares_outstanding: z.number().nullable(),
  diluted_shares_outstanding: z.number().nullable(),
  ebitda: z.number().nullable(),
  net_income_continuous_operations: z.number().nullable(),
  minority_interests: z.number().nullable(),
  preferred_stock_dividends: z.number().nullable(),

  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(), // ISO datetime string from backend
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new income statement (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class IncomeStatementCreate(BaseModel)
 */
export const CreateIncomeStatementSchema = z.object({
  entity_id: z.number(),
  year: z.number(),
  full_year: z.boolean().optional(),
  scenario: z.string().optional(),
  quarter: z.string().nullable().optional(),
  semester: z.string().nullable().optional(),
  month: z.string().nullable().optional(),
  period_start: z.string().nullable().optional(),
  period_end: z.string().nullable().optional(),

  // Revenue
  revenue: z.number().nullable().optional(),
  cost_of_goods: z.number().nullable().optional(),
  gross_profit: z.number().nullable().optional(),

  // Operating Expenses
  research_and_development: z.number().nullable().optional(),
  selling_general_and_administrative: z.number().nullable().optional(),
  other_operating_expenses: z.number().nullable().optional(),

  // Results
  operating_income: z.number().nullable().optional(),
  non_operating_interest_income: z.number().nullable().optional(),
  non_operating_interest_expense: z.number().nullable().optional(),
  other_income_expense: z.number().nullable().optional(),
  pretax_income: z.number().nullable().optional(),
  income_tax: z.number().nullable().optional(),
  net_income: z.number().nullable().optional(),

  // Additional
  eps_basic: z.number().nullable().optional(),
  eps_diluted: z.number().nullable().optional(),
  basic_shares_outstanding: z.number().nullable().optional(),
  diluted_shares_outstanding: z.number().nullable().optional(),
  ebitda: z.number().nullable().optional(),
  net_income_continuous_operations: z.number().nullable().optional(),
  minority_interests: z.number().nullable().optional(),
  preferred_stock_dividends: z.number().nullable().optional(),
});

/**
 * Schema for updating an income statement (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class IncomeStatementUpdate(BaseModel)
 */
export const UpdateIncomeStatementSchema = z.object({
  entity_id: z.number().nullable().optional(),

  // Time Dimensions
  year: z.number().nullable().optional(),
  quarter: z.string().nullable().optional(),
  semester: z.string().nullable().optional(),
  month: z.string().nullable().optional(),
  full_year: z.boolean().nullable().optional(),
  scenario: z.string().nullable().optional(),
  period_start: z.string().nullable().optional(),
  period_end: z.string().nullable().optional(),

  // Revenue
  revenue: z.number().nullable().optional(),
  cost_of_goods: z.number().nullable().optional(),
  gross_profit: z.number().nullable().optional(),

  // Operating Expenses
  research_and_development: z.number().nullable().optional(),
  selling_general_and_administrative: z.number().nullable().optional(),
  other_operating_expenses: z.number().nullable().optional(),

  // Results
  operating_income: z.number().nullable().optional(),
  non_operating_interest_income: z.number().nullable().optional(),
  non_operating_interest_expense: z.number().nullable().optional(),
  other_income_expense: z.number().nullable().optional(),
  pretax_income: z.number().nullable().optional(),
  income_tax: z.number().nullable().optional(),
  net_income: z.number().nullable().optional(),

  // Additional
  eps_basic: z.number().nullable().optional(),
  eps_diluted: z.number().nullable().optional(),
  basic_shares_outstanding: z.number().nullable().optional(),
  diluted_shares_outstanding: z.number().nullable().optional(),
  ebitda: z.number().nullable().optional(),
  net_income_continuous_operations: z.number().nullable().optional(),
  minority_interests: z.number().nullable().optional(),
  preferred_stock_dividends: z.number().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type IncomeStatement = z.infer<typeof IncomeStatementSchema>;
export type CreateIncomeStatement = z.infer<typeof CreateIncomeStatementSchema>;
export type UpdateIncomeStatement = z.infer<typeof UpdateIncomeStatementSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single income statement
 * Backend equivalent: class IncomeStatementResponse(BaseModel)
 */
export type IncomeStatementResponse = {
  success: boolean;
  data?: IncomeStatement;
  error?: string;
};

/**
 * Response containing multiple income statements
 * Backend equivalent: class IncomeStatementsResponse(BaseModel)
 */
export type IncomeStatementsResponse = {
  success: boolean;
  data?: IncomeStatement[];
  error?: string;
};

File: server/apps/assetmanager/schemas/financial_schemas/income_statement_schemas.py
"""
IncomeStatement Schemas

Pydantic schemas for the IncomeStatement model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type

# ==========================================
# IncomeStatement Schema (Full Representation)
# ==========================================

class IncomeStatement(BaseModel):
    """IncomeStatement schema - full representation"""
    id: int
    entity_id: int = Field(description="Associated entity ID")

    # Time Dimensions
    year: int = Field(description="Fiscal year")
    quarter: str | None = Field(None, description="Fiscal quarter")
    semester: str | None = Field(None, description="Fiscal semester")
    month: str | None = Field(None, description="Fiscal month")
    full_year: bool = Field(description="Whether this is a full year statement")
    scenario: str = Field(description="Scenario type (actual, budget, forecast)")
    period_start: date_type | None = Field(None, description="Period start date")
    period_end: date_type | None = Field(None, description="Period end date")

    # Revenue
    revenue: float | None = Field(None, description="Total revenue")
    cost_of_goods: float | None = Field(None, description="Cost of goods sold")
    gross_profit: float | None = Field(None, description="Gross profit")

    # Operating Expenses
    research_and_development: float | None = Field(None, description="Research and development expenses")
    selling_general_and_administrative: float | None = Field(None, description="Selling, general and administrative expenses")
    other_operating_expenses: float | None = Field(None, description="Other operating expenses")

    # Results
    operating_income: float | None = Field(None, description="Operating income")
    non_operating_interest_income: float | None = Field(None, description="Non-operating interest income")
    non_operating_interest_expense: float | None = Field(None, description="Non-operating interest expense")
    other_income_expense: float | None = Field(None, description="Other income/expense")
    pretax_income: float | None = Field(None, description="Pretax income")
    income_tax: float | None = Field(None, description="Income tax")
    net_income: float | None = Field(None, description="Net income")

    # Additional
    eps_basic: float | None = Field(None, description="Basic earnings per share")
    eps_diluted: float | None = Field(None, description="Diluted earnings per share")
    basic_shares_outstanding: float | None = Field(None, description="Basic shares outstanding")
    diluted_shares_outstanding: float | None = Field(None, description="Diluted shares outstanding")
    ebitda: float | None = Field(None, description="EBITDA")
    net_income_continuous_operations: float | None = Field(None, description="Net income from continuous operations")
    minority_interests: float | None = Field(None, description="Minority interests")
    preferred_stock_dividends: float | None = Field(None, description="Preferred stock dividends")

    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class IncomeStatementCreate(BaseModel):
    """Schema for creating a new income statement"""
    # Required fields
    entity_id: int = Field(description="Associated entity ID")
    year: int = Field(description="Fiscal year")

    # Fields with defaults
    full_year: bool = Field(default=False, description="Whether this is a full year statement")
    scenario: str = Field(default="actual", description="Scenario type (actual, budget, forecast)")

    # Optional fields
    quarter: str | None = Field(None, description="Fiscal quarter")
    semester: str | None = Field(None, description="Fiscal semester")
    month: str | None = Field(None, description="Fiscal month")
    period_start: date_type | None = Field(None, description="Period start date")
    period_end: date_type | None = Field(None, description="Period end date")

    # Revenue
    revenue: float | None = Field(None, description="Total revenue")
    cost_of_goods: float | None = Field(None, description="Cost of goods sold")
    gross_profit: float | None = Field(None, description="Gross profit")

    # Operating Expenses
    research_and_development: float | None = Field(None, description="Research and development expenses")
    selling_general_and_administrative: float | None = Field(None, description="Selling, general and administrative expenses")
    other_operating_expenses: float | None = Field(None, description="Other operating expenses")

    # Results
    operating_income: float | None = Field(None, description="Operating income")
    non_operating_interest_income: float | None = Field(None, description="Non-operating interest income")
    non_operating_interest_expense: float | None = Field(None, description="Non-operating interest expense")
    other_income_expense: float | None = Field(None, description="Other income/expense")
    pretax_income: float | None = Field(None, description="Pretax income")
    income_tax: float | None = Field(None, description="Income tax")
    net_income: float | None = Field(None, description="Net income")

    # Additional
    eps_basic: float | None = Field(None, description="Basic earnings per share")
    eps_diluted: float | None = Field(None, description="Diluted earnings per share")
    basic_shares_outstanding: float | None = Field(None, description="Basic shares outstanding")
    diluted_shares_outstanding: float | None = Field(None, description="Diluted shares outstanding")
    ebitda: float | None = Field(None, description="EBITDA")
    net_income_continuous_operations: float | None = Field(None, description="Net income from continuous operations")
    minority_interests: float | None = Field(None, description="Minority interests")
    preferred_stock_dividends: float | None = Field(None, description="Preferred stock dividends")

class IncomeStatementUpdate(BaseModel):
    """Schema for updating an income statement"""
    entity_id: int | None = None

    # Time Dimensions
    year: int | None = None
    quarter: str | None = None
    semester: str | None = None
    month: str | None = None
    full_year: bool | None = None
    scenario: str | None = None
    period_start: date_type | None = None
    period_end: date_type | None = None

    # Revenue
    revenue: float | None = None
    cost_of_goods: float | None = None
    gross_profit: float | None = None

    # Operating Expenses
    research_and_development: float | None = None
    selling_general_and_administrative: float | None = None
    other_operating_expenses: float | None = None

    # Results
    operating_income: float | None = None
    non_operating_interest_income: float | None = None
    non_operating_interest_expense: float | None = None
    other_income_expense: float | None = None
    pretax_income: float | None = None
    income_tax: float | None = None
    net_income: float | None = None

    # Additional
    eps_basic: float | None = None
    eps_diluted: float | None = None
    basic_shares_outstanding: float | None = None
    diluted_shares_outstanding: float | None = None
    ebitda: float | None = None
    net_income_continuous_operations: float | None = None
    minority_interests: float | None = None
    preferred_stock_dividends: float | None = None

# ==========================================
# Response Types
# ==========================================

class IncomeStatementResponse(BaseModel):
    """Response containing a single income statement"""
    success: bool
    data: IncomeStatement | None = None
    error: str | None = None

class IncomeStatementsResponse(BaseModel):
    """Response containing multiple income statements"""
    success: bool
    data: list[IncomeStatement] | None = None
    error: str | None = None
